File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
from sys import stdin
from itertools import permutations

N = int(stdin.readline().strip())
nums = list(map(int, stdin.readline().split()))
ops = list(map(int, stdin.readline().split())) # +, -, *, // 개수

max_result = -float('inf') # 연산 최댓값
min_result = float('inf') # 연산 최솟값

def dfs(cnt, curr_result, plus, minus, multiply, divide):
    global max_result, min_result
    
    # 마지막 숫자까지 연산 완료
    if cnt == N:
        max_result = max(max_result, curr_result)
        min_result = min(min_result, curr_result)
        return
    
    if plus:
        dfs(cnt+1, curr_result + nums[cnt], plus-1, minus, multiply, divide)
    if minus:
        dfs(cnt+1, curr_result - nums[cnt], plus, minus-1, multiply, divide)
    if multiply:
        dfs(cnt+1, curr_result * nums[cnt], plus, minus, multiply-1, divide)
    if divide:
        if curr_result < 0:
            dfs(cnt+1, -(-curr_result // nums[cnt]), plus, minus, multiply, divide-1)
        else:
            dfs(cnt+1, curr_result // nums[cnt], plus, minus, multiply, divide-1)

dfs(1, nums[0], ops[0], ops[1], ops[2], ops[3])

print(max_result)
print(min_result)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static int[] nums;
    static int[] ops = new int[4]; // + - * /
    static int maxResult = Integer.MIN_VALUE;;
    static int minResult = Integer.MAX_VALUE;;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        nums = new int[N];
        for (int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 4; i++) {
            ops[i] = Integer.parseInt(st.nextToken());
        }

        dfs(1, nums[0], ops[0], ops[1], ops[2], ops[3]);

        System.out.println(maxResult);
        System.out.println(minResult);
    }

    static void dfs(int cnt, int currResult, int plus, int minus, int multiply, int divide) {
        if (cnt == N) {
            maxResult = Math.max(maxResult, currResult);
            minResult = Math.min(minResult, currResult);
            return;
        }

        if (plus > 0) { // 더하기 연산자가 더 남아있다면
            dfs(cnt + 1, currResult + nums[cnt], plus - 1, minus, multiply, divide);
        }
        if (minus > 0) {
            dfs(cnt + 1, currResult - nums[cnt], plus, minus - 1, multiply, divide);
        }
        if (multiply > 0) {
            dfs(cnt + 1, currResult * nums[cnt], plus, minus, multiply - 1, divide);
        }
        if (divide > 0) {
            int temp;
            if (currResult < 0) {
                temp = -(-currResult / nums[cnt]);
            } else {
                temp = currResult / nums[cnt];
            }
            dfs(cnt + 1, temp, plus, minus, multiply, divide - 1);
        }
    }
}